### ISSUE TITLE
GRPC modules fail to compile on RHEL 9

### ISSUE THREAD

**Issue by czanik** [syslog-ng/syslog-ng#4654]
I tried to compile GRPC modules again on RHEL 9 (well, actually Rocky 9). I used 4.4.0 release sources. Compilation fails with the following error:
```
  CXX      modules/grpc/otel/libotel_cpp_la-syslog-ng-otlp-dest-worker.lo
  CXX      modules/grpc/credentials/loki_libloki_cpp_la-grpc-credentials-builder.lo
  CXX      modules/grpc/loki/libloki_cpp_la-loki-dest.lo
In file included from /usr/lib64/glib-2.0/include/glibconfig.h:9,
                 from /usr/include/glib-2.0/glib/gtypes.h:32,
                 from /usr/include/glib-2.0/glib/galloca.h:32,
                 from /usr/include/glib-2.0/glib.h:30,
                 from ./lib/compat/glib.h:29,
                 from ./lib/syslog-ng.h:35,
                 from modules/grpc/loki/loki-dest.h:28,
                 from modules/grpc/loki/loki-dest.hpp:26,
                 from modules/grpc/loki/loki-dest.cpp:23:
./lib/atomic-gssize.h: In function 'gboolean atomic_gssize_compare_and_exchange(atomic_gssize*, gssize, gssize)':
/usr/include/glib-2.0/glib/gmacros.h:247:38: error: 'remove_reference' in namespace 'std' does not name a template type
  247 | #define glib_typeof(t) typename std::remove_reference<decltype (t)>::type
      |                                      ^~~~~~~~~~~~~~~~
/usr/include/glib-2.0/glib/gatomic.h:203:5: note: in expansion of macro 'glib_typeof'
  203 |     glib_typeof (*(atomic)) gapcae_oldval = (oldval);                        \
      |     ^~~~~~~~~~~
./lib/atomic-gssize.h:118:10: note: in expansion of macro 'g_atomic_pointer_compare_and_exchange'
  118 |   return g_atomic_pointer_compare_and_exchange(&a->value, oldval, newval);
      |          ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/usr/include/glib-2.0/glib/gmacros.h:247:54: error: expected unqualified-id before '<' token
  247 | #define glib_typeof(t) typename std::remove_reference<decltype (t)>::type
      |                                                      ^
/usr/include/glib-2.0/glib/gatomic.h:203:5: note: in expansion of macro 'glib_typeof'
  203 |     glib_typeof (*(atomic)) gapcae_oldval = (oldval);                        \
      |     ^~~~~~~~~~~
./lib/atomic-gssize.h:118:10: note: in expansion of macro 'g_atomic_pointer_compare_and_exchange'
  118 |   return g_atomic_pointer_compare_and_exchange(&a->value, oldval, newval);
      |          ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
In file included from /usr/include/glib-2.0/glib/gthread.h:32,
                 from /usr/include/glib-2.0/glib/gasyncqueue.h:32,
                 from /usr/include/glib-2.0/glib.h:32,
                 from ./lib/compat/glib.h:29,
                 from ./lib/syslog-ng.h:35,
                 from modules/grpc/loki/loki-dest.h:28,
                 from modules/grpc/loki/loki-dest.hpp:26,
                 from modules/grpc/loki/loki-dest.cpp:23:
/usr/include/glib-2.0/glib/gatomic.h:206:45: error: 'gapcae_oldval' was not declared in this scope
  206 |     __atomic_compare_exchange_n ((atomic), &gapcae_oldval, (newval), FALSE, __ATOMIC_SEQ_CST, __ATOMIC_SEQ_CST) ? TRUE : FALSE; \
      |                                             ^~~~~~~~~~~~~
./lib/atomic-gssize.h:118:10: note: in expansion of macro 'g_atomic_pointer_compare_and_exchange'
  118 |   return g_atomic_pointer_compare_and_exchange(&a->value, oldval, newval);
      |          ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/usr/include/glib-2.0/glib/gatomic.h:201:21: error: void value not ignored as it ought to be
  201 |   (G_GNUC_EXTENSION ({                                                       \
      |   ~~~~~~~~~~~~~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
  202 |     G_STATIC_ASSERT (sizeof (oldval) == sizeof (gpointer));                  \
      |     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
  203 |     glib_typeof (*(atomic)) gapcae_oldval = (oldval);                        \
      |     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
  204 |     G_STATIC_ASSERT (sizeof *(atomic) == sizeof (gpointer));                 \
      |     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
  205 |     (void) (0 ? (gpointer) *(atomic) : NULL);                                \
      |     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
  206 |     __atomic_compare_exchange_n ((atomic), &gapcae_oldval, (newval), FALSE, __ATOMIC_SEQ_CST, __ATOMIC_SEQ_CST) ? TRUE : FALSE; \
      |     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
  207 |   }))
      |   ~~~                
./lib/atomic-gssize.h:118:10: note: in expansion of macro 'g_atomic_pointer_compare_and_exchange'
  118 |   return g_atomic_pointer_compare_and_exchange(&a->value, oldval, newval);
      |          ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
  CXX      modules/grpc/loki/libloki_cpp_la-loki-worker.lo
In file included from /usr/lib64/glib-2.0/include/glibconfig.h:9,
                 from /usr/include/glib-2.0/glib/gtypes.h:32,
                 from /usr/include/glib-2.0/glib/galloca.h:32,
                 from /usr/include/glib-2.0/glib.h:30,
                 from ./lib/compat/glib.h:29,
                 from ./lib/syslog-ng.h:35,
                 from modules/grpc/loki/loki-worker.h:28,
                 from modules/grpc/loki/loki-worker.hpp:26,
                 from modules/grpc/loki/loki-worker.cpp:23:
./lib/atomic-gssize.h: In function 'gboolean atomic_gssize_compare_and_exchange(atomic_gssize*, gssize, gssize)':
/usr/include/glib-2.0/glib/gmacros.h:247:38: error: 'remove_reference' in namespace 'std' does not name a template type
  247 | #define glib_typeof(t) typename std::remove_reference<decltype (t)>::type
      |                                      ^~~~~~~~~~~~~~~~
/usr/include/glib-2.0/glib/gatomic.h:203:5: note: in expansion of macro 'glib_typeof'
  203 |     glib_typeof (*(atomic)) gapcae_oldval = (oldval);                        \
      |     ^~~~~~~~~~~
./lib/atomic-gssize.h:118:10: note: in expansion of macro 'g_atomic_pointer_compare_and_exchange'
  118 |   return g_atomic_pointer_compare_and_exchange(&a->value, oldval, newval);
      |          ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/usr/include/glib-2.0/glib/gmacros.h:247:54: error: expected unqualified-id before '<' token
  247 | #define glib_typeof(t) typename std::remove_reference<decltype (t)>::type
      |                                                      ^
/usr/include/glib-2.0/glib/gatomic.h:203:5: note: in expansion of macro 'glib_typeof'
  203 |     glib_typeof (*(atomic)) gapcae_oldval = (oldval);                        \
      |     ^~~~~~~~~~~
./lib/atomic-gssize.h:118:10: note: in expansion of macro 'g_atomic_pointer_compare_and_exchange'
  118 |   return g_atomic_pointer_compare_and_exchange(&a->value, oldval, newval);
      |          ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
In file included from /usr/include/glib-2.0/glib/gthread.h:32,
                 from /usr/include/glib-2.0/glib/gasyncqueue.h:32,
                 from /usr/include/glib-2.0/glib.h:32,
                 from ./lib/compat/glib.h:29,
                 from ./lib/syslog-ng.h:35,
                 from modules/grpc/loki/loki-worker.h:28,
                 from modules/grpc/loki/loki-worker.hpp:26,
                 from modules/grpc/loki/loki-worker.cpp:23:
/usr/include/glib-2.0/glib/gatomic.h:206:45: error: 'gapcae_oldval' was not declared in this scope
  206 |     __atomic_compare_exchange_n ((atomic), &gapcae_oldval, (newval), FALSE, __ATOMIC_SEQ_CST, __ATOMIC_SEQ_CST) ? TRUE : FALSE; \
      |                                             ^~~~~~~~~~~~~
./lib/atomic-gssize.h:118:10: note: in expansion of macro 'g_atomic_pointer_compare_and_exchange'
  118 |   return g_atomic_pointer_compare_and_exchange(&a->value, oldval, newval);
      |          ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/usr/include/glib-2.0/glib/gatomic.h:201:21: error: void value not ignored as it ought to be
  201 |   (G_GNUC_EXTENSION ({                                                       \
      |   ~~~~~~~~~~~~~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
  202 |     G_STATIC_ASSERT (sizeof (oldval) == sizeof (gpointer));                  \
      |     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
  203 |     glib_typeof (*(atomic)) gapcae_oldval = (oldval);                        \
      |     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
  204 |     G_STATIC_ASSERT (sizeof *(atomic) == sizeof (gpointer));                 \
      |     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
  205 |     (void) (0 ? (gpointer) *(atomic) : NULL);                                \
      |     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
  206 |     __atomic_compare_exchange_n ((atomic), &gapcae_oldval, (newval), FALSE, __ATOMIC_SEQ_CST, __ATOMIC_SEQ_CST) ? TRUE : FALSE; \
      |     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
  207 |   }))
      |   ~~~                
./lib/atomic-gssize.h:118:10: note: in expansion of macro 'g_atomic_pointer_compare_and_exchange'
  118 |   return g_atomic_pointer_compare_and_exchange(&a->value, oldval, newval);
      |          ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
make[2]: *** [Makefile:26746: modules/grpc/loki/libloki_cpp_la-loki-dest.lo] Error 1
make[2]: *** Waiting for unfinished jobs....
make[2]: *** [Makefile:26753: modules/grpc/loki/libloki_cpp_la-loki-worker.lo] Error 1
make[1]: *** [Makefile:28067: all-recursive] Error 1
make: *** [Makefile:12058: all] Error 2
```
Summary at the end of ./configure:
```
syslog-ng Open Source Edition 4.4.0 configured
 Edition settings:
  Release type                : stable
  Pretty version              : 4
  Combined vers.              : 4 (4.4.0)
  Package name                : syslog-ng
 Compiler options:
  compiler                    : gcc
  compiler options            :  -O2 -g -pthread -O2 -flto=auto -ffat-lto-objects -fexceptions -g -grecord-gcc-switches -pipe -Wall -Werror=format-security -Wp,-D_FORTIFY_SOURCE=2 -Wp,-D_GLIBCXX_ASSERTIONS -specs=/usr/lib/rpm/redhat/redhat-hardened-cc1 -fstack-protector-strong -specs=/usr/lib/rpm/redhat/redhat-annobin-cc1  -m64 -march=x86-64-v2 -mtune=generic -fasynchronous-unwind-tables -fstack-clash-protection -fcf-protection  -I/usr/include/glib-2.0 -I/usr/lib64/glib-2.0/include -I/usr/include/sysprof-4 -pthread  -DGLIB_VERSION_MIN_REQUIRED=GLIB_VERSION_2_32 -I$(top_srcdir)/lib/eventlog/src -I$(top_builddir)/lib/eventlog/src   -D_BSD_SOURCE -D_DEFAULT_SOURCE -DHAVE_NET_ETHERNET_H -D_DEFAULT_SOURCE    -D_GNU_SOURCE -D_DEFAULT_SOURCE -D_LARGEFILE_SOURCE -D_FILE_OFFSET_BITS=64
  C++ enabled                 : yes
  C++ compiler                : g++
  C++ compiler options        :  -O2 -g -O2 -flto=auto -ffat-lto-objects -fexceptions -g -grecord-gcc-switches -pipe -Wall -Werror=format-security -Wp,-D_FORTIFY_SOURCE=2 -Wp,-D_GLIBCXX_ASSERTIONS -specs=/usr/lib/rpm/redhat/redhat-hardened-cc1 -fstack-protector-strong -specs=/usr/lib/rpm/redhat/redhat-annobin-cc1  -m64 -march=x86-64-v2 -mtune=generic -fasynchronous-unwind-tables -fstack-clash-protection -fcf-protection
  linker flags                : -Wl,-z,relro -Wl,--as-needed  -Wl,-z,now -specs=/usr/lib/rpm/redhat/redhat-hardened-ld -specs=/usr/lib/rpm/redhat/redhat-annobin-cc1  
  prefix                      : /usr
  linking mode                : dynamic
  embedded crypto             : yes
  __thread keyword            : yes
 Test dependencies:
  Criterion                   : no
  Unit tests                  : no
 Submodules:
  ivykis                      : system
 Features:
  Forced server mode          : yes
  Debug symbols               : no
  GCC profiling               : no
  Memtrace                    : no
  IPV6 support                : yes
  spoof-source support        : yes
  tcp-wrapper support         : no
  Linux capability support    : yes
  Env wrapper support         : no
  systemd support             : yes (unit dir: /usr/lib/systemd/system)
  systemd-journal support     : system
  JSON support                : yes
 Build options:
  Generate manual pages       : yes
  Install manual pages        : yes
 Modules:
  Module search path          : /usr/lib64/syslog-ng
  Sun STREAMS support (module): no
  OpenBSD syslog (module)     : no
  SQL support (module)        : auto
  PACCT module (EXPERIMENTAL) : yes
  MongoDB destination (module): yes
  JSON support (module)       : yes
  SMTP support (module)       : yes
  HTTP support (module)       : yes
  AMQP destination (module)   : yes
  STOMP destination (module)  : yes
  MQTT (module)               : no
  GRPC (module)               : yes
  GEOIP2 support (module)     : yes
  Redis support (module)      : yes
  Kafka support (module)      : yes
  Riemann destination (module): yes, microseconds: no
  Python                      : yes (pkg-config package: python-3.9, packages: none)
  Python modules              : yes
  Java                        : yes
  Java modules                : no
  afsnmp module               : yes
  eBPF module                 : no
  native bindings             : auto
```

**Comment by alltilla** [comment]
This is probably a glib bug. I've seen similar here: https://bugs.launchpad.net/ubuntu/+source/qemu/+bug/1916705, but the fixed version is available in rocky 9, so either this is a different issue, or the fix was not complete.

**Comment by czanik** [comment]
The c++ part of the examples module fails to compile with similar error messages on RHEL 9.

**Comment by bazsi** [comment]
This should be fixed by #4739 once that becomes green. Also I am not sure if we could compile syslog-ng to rhel9 as a part of its CI.

We used to use CentOS for this purpose but for now I don't know what we could do to mimic rhel9 without actually buying a subscription. How do you compile your packages?

**Comment by czanik** [comment]
Copr uses RHEL 9 + EPEL

For local builds I use the ```mock``` command on Fedora, and uses Rocky Linux + EPEL.
